mlflow/models/utils.py

```diff
@@ -462,6 +462,54 @@ def _enforce_col_schema(pf_input: PyFuncInput, input_schema: Schema):
     return new_pf_input
 
 
+def _reshape_and_cast_pandas_column_values(name, pd_series, tensor_spec):
+    if tensor_spec.shape[0] != -1 or -1 in tensor_spec.shape[1:]:
+        raise MlflowException(
+            "For pandas dataframe input, the first dimension of shape must be a variable "
+            "dimension and other dimensions must be fixed, but in model signature the shape "
+            f"of {'input ' + name if name else 'the unnamed input'} is {tensor_spec.shape}."
+        )
+
+    if np.isscalar(pd_series[0]):
+        if tensor_spec.shape != (-1,):
+            raise MlflowException(
+                f"The input pandas dataframe column '{name}' contains scalar "
+                "values, which requires the shape to be (-1,), but got tensor spec "
+                f"shape of {tensor_spec.shape}.",
+                error_code=INVALID_PARAMETER_VALUE,
+            )
+        return _enforce_tensor_spec(np.array(pd_series, dtype=tensor_spec.type), tensor_spec)
+    elif isinstance(pd_series[0], list) and np.isscalar(pd_series[0][0]):
+        # If the pandas column contains list type values,
+        # in this case, the shape and type information is lost,
+        # so do not enforce the shape and type, instead,
+        # reshape the array value list to the required shape, and cast value type to
+        # required type.
+        err_msg = (
+            f"The value in the Input DataFrame column '{name}' could not be converted to the "
+            f"expected shape of: '{tensor_spec.shape}'. Ensure that each of the input list "
+            "elements are of uniform length and that the data can be coerced to the tensor "
+            f"type '{tensor_spec.type}'"
+        )
+        try:
+            flattened_numpy_arr = np.vstack(pd_series.tolist())
+            reshaped_numpy_arr = flattened_numpy_arr.reshape(tensor_spec.shape).astype(
+                tensor_spec.type
+            )
+        except ValueError:
+            raise MlflowException(err_msg, error_code=INVALID_PARAMETER_VALUE)
+        if len(reshaped_numpy_arr) != len(pd_series):
+            raise MlflowException(err_msg, error_code=INVALID_PARAMETER_VALUE)
+        return reshaped_numpy_arr
+    else:
+        raise MlflowException(
+            "Because the model signature requires tensor spec input, the input "
```

Comment:
Can we also support the case when the pandas dataframe can directly be converted into expected shape, i.e. the if the signature is a dict of 1d tensors or if it is a single 2d tensor with shape matches pdf.values.shape. 

Alternatively, can we just enable this per spark_udf instance by adding an argument to mlflow.spark_udf call? E.g. reshape_tensor_input=False.

Reply:
> a single 2d tensor with shape matches pdf.values.shape.

This is supported. See this line https://github.com/mlflow/mlflow/pull/7291/files#diff-395a8aeaf3442dfb27e45ec1a038b71c76c7b8ad42744ee67587c8314acfee04R574

> the signature is a dict of 1d tensors

This is also supported, it requires all pandas column values to be a list and skip reshaping. This line already can handle the case: https://github.com/mlflow/mlflow/pull/7291/files#diff-395a8aeaf3442dfb27e45ec1a038b71c76c7b8ad42744ee67587c8314acfee04R540

Reply:
Oh I missed that in the code, thanks for the clarification!